### ISSUE TITLE
Enhancement: Input invalid state styling

### ISSUE THREAD

**Issue by capeoples** [Esri/calcite-components#1592]
### Description
Currently, input invalid states have the label turn red, and will have the error message below.
<img width="454" alt="Screen Shot 2021-02-22 at 9 49 52 AM" src="https://user-images.githubusercontent.com/30473086/108748684-b8215800-74f3-11eb-9388-cf0c5353abb2.png">

This can create a weird red/blue state when the input is in focus:
<img width="432" alt="Screen Shot 2021-02-22 at 9 54 18 AM" src="https://user-images.githubusercontent.com/30473086/108748873-f3238b80-74f3-11eb-960d-af0e16d3e469.png">

To address this, I'd like to propose that the red color highlight happens on the input field instead of the label. 
<img width="918" alt="Screen Shot 2021-02-22 at 9 56 07 AM" src="https://user-images.githubusercontent.com/30473086/108749073-341ba000-74f4-11eb-9d37-e1a0a9615787.png">

#### Which Component
Input. Specifically looking at input with label and input message, but I supposed this could apply to all input types. 

I would note, that I'm not sure how the input without label displays the error state. 

#### Next steps
@macandcheese, can you notate the additional components you think should be considered with this update so @abigailmbravo-uxuidev can do design exploration on this?

cc @bstifle, @julio8a, @paulcpederson

**Comment by paulcpederson** [comment]
I agree this is a good change. Surveying the current components that van have invalid or red states they pretty uniformly use border/background/icon to communicate it:

<img width="242" alt="Screen Shot 2021-02-22 at 2 19 23 PM" src="https://user-images.githubusercontent.com/1031758/108777739-22002880-7519-11eb-8ed7-90f1faccdfaf.png">
<img width="186" alt="Screen Shot 2021-02-22 at 2 18 42 PM" src="https://user-images.githubusercontent.com/1031758/108777741-2298bf00-7519-11eb-8e70-b666d40ed74b.png">
<img width="649" alt="Screen Shot 2021-02-22 at 2 18 57 PM" src="https://user-images.githubusercontent.com/1031758/108777743-23315580-7519-11eb-8777-55474e21e5e4.png">
<img width="228" alt="Screen Shot 2021-02-22 at 2 18 35 PM" src="https://user-images.githubusercontent.com/1031758/108777744-23315580-7519-11eb-9b6f-6204da9c3529.png">
<img width="572" alt="Screen Shot 2021-02-22 at 2 17 55 PM" src="https://user-images.githubusercontent.com/1031758/108777745-23c9ec00-7519-11eb-81bf-90760736097f.png">

Then you look at inputs which currently use just text color, it feels like an outlier for sure:

<img width="462" alt="Screen Shot 2021-02-22 at 2 18 07 PM" src="https://user-images.githubusercontent.com/1031758/108777792-36dcbc00-7519-11eb-8fe3-6ca6287c24ad.png">
<img width="436" alt="Screen Shot 2021-02-22 at 2 18 15 PM" src="https://user-images.githubusercontent.com/1031758/108777793-380de900-7519-11eb-820e-f7e473762e67.png">

**Comment by bstifle** [comment]
Awesome. @capeoples looking forward to seeing that exhaustive form view!!

**Comment by ethanbdev** [comment]
👍 for this as well! We are currently overriding the blue border with red, so this would be great to remove that.

**Comment by abigailmbravo-uxuidev** [comment]
Hey everyone! Here is a list of the invalid error states around form controls. For consistency, @capeoples and I wanted to explore removing the red label on input and this is how the rest of the form control would be represented. We have discussed with @macandcheese and we have all agreed on what we have here! 

This is ready to go. Please let me know if anyone has any questions. 
<img width="837" alt="Frame 2" src="https://user-images.githubusercontent.com/73944682/109888784-f7108580-7c38-11eb-9bd3-c02b6211acbc.png">

**Comment by TheBlueDog** [comment]
Because in many cases the validation message is critical to our users workflow, we would prefer to see that text in red also to provide additional visual distinction from the description text.

<img width="383" alt="Screen Shot 2021-03-16 at 9 20 17 PM" src="https://user-images.githubusercontent.com/5419306/111400383-9acc4d80-869d-11eb-950c-ffcd38260cf1.png">

**Comment by paulcpederson** [comment]
I just checked and the ui-danger passes WCAG 2.0 on the ui-background, so would be fine from an accessibility perspective.

**Comment by bstifle** [comment]
Green, and yellow do not pass. so we'd be inconsistent by doing that

**Comment by bstifle** [comment]
also, does not pass on all foregrounds

![image](https://user-images.githubusercontent.com/14083530/111654255-95bee980-87d6-11eb-84f0-bc826f7fd864.png)

![image](https://user-images.githubusercontent.com/14083530/111654302-a0797e80-87d6-11eb-9865-f9f897404e35.png)


i'd call that a no-go

**Comment by bstifle** [comment]
let's move forward with the red input border, and removing the red input-label styling.

**Comment by paulcpederson** [comment]
Cool I can try to help update these.

**Comment by TheBlueDog** [comment]
As long as the color is not in the shadow can be set or overridden by the app

**Comment by bstifle** [comment]
@nick-esri can you post a screenshot please

**Comment by nick-esri** [comment]
Since Description, Helper Text and Error Message are presented beneath an input, they are mutually exclusive:
- **Description**: A string placed beneath an input that provides instructions as to how to fill out the field. Descriptions are optional and restricted to a maximum of 255 characters.
- **Helper Text**: A string placed beneath an input that provides character constraint information for strings, or range domain constraints for numeric fields.
- **Error Message**: A string placed beneath an input when in an error state, that informs the user about an invalid value.

Rules:
- When a **Description** is specified, it is always displayed except when the field is in an error state, in which case it is replaced by the **Error Message**.
- When a **Description** is not specified, the **Helper Text** is displayed when the field is focused. If it is in error state, it is replaced by the **Error Message**.


Here's some examples from iOS:

### String Field - Character Constraints
![image](https://user-images.githubusercontent.com/3908500/112677871-9ee83000-8e40-11eb-8904-1b5d36b71de0.png)


### String Field - Description
![image](https://user-images.githubusercontent.com/3908500/112677923-af000f80-8e40-11eb-8341-c9539d573702.png)

### Numeric Field - Range Domain
![image](https://user-images.githubusercontent.com/3908500/112678083-e5d62580-8e40-11eb-8308-41c8bbd2d469.png)

### Numeric Field - Description
![image](https://user-images.githubusercontent.com/3908500/112678214-08683e80-8e41-11eb-8377-7a9eef3da738.png)

**Comment by bstifle** [comment]
![image](https://user-images.githubusercontent.com/14083530/113352375-c5571100-9301-11eb-8441-24b1102dca8e.png)

so if we use ui-danger-hover, it passes on all 3 foreground colors.

I'm not convinced this is an improvement

**Comment by TheBlueDog** [comment]
This is using `v-rr-180` (`#B3000F`) on top, as compared to `h-rr-070` below. It passes AA on all. Shown here against `ui-foreground-3`

![Screen Shot 2021-04-01 at 6 07 53 PM](https://user-images.githubusercontent.com/5419306/113359164-3653f400-9315-11eb-95d0-c36485889f50.png)

**Comment by bstifle** [comment]
Right, yeah that color is about the same as #A82B1E visually.

We could either:
- keep existing
- use ui-danger-hover for these error states
- or rework all of the reds on light theme to be darker

**Comment by nick-esri** [comment]
If the text is being tinted we probably don’t need the border of the field to be red- feels like overkill.

Tangentially related, can I ask about the size of the icon? It looks a little large compared to the font size.  

Agree with Bryan on the super red box- not a fan either.

**Comment by TheBlueDog** [comment]
And the red border get hidden by the focus ring. Unless we can change the focus ring color on fields with errors to red.

**Comment by nick-esri** [comment]
@bstifle - I know we probably need to drive to a decision here. 

> We could either:
> - keep existing
> - use `ui-danger-hover` for these error states
> - or rework all of the reds on light theme to be darker

On reflection I think it's reasonable to "keep existing" if that means going with with the red border / red focus ring when the field is in error. Can you confirm that when the field is in focus the border will be red too?

I quite like some other aspects of @capeoples proposal in the description:
![](https://user-images.githubusercontent.com/30473086/108749073-341ba000-74f4-11eb-9d37-e1a0a9615787.png)

Namely the error text is a little smaller and the error text looks to be gray. Might be out of scope for this particular issue.

**Comment by TheBlueDog** [comment]
Tailwind has a built in solution for handling the focus ring, color, width, inset, opacity and offset. Seems like that could be really useful here. https://tailwindcss.com/docs/ring-color

**Comment by bstifle** [comment]
@nick-esri 

yeah I think making the text gray could be nice.

Yes, red error borders are gonna be in for sure. As far as focusing then and staying red, i think that makes sense for most cases, but i believe we'll have a way to disable that.

**Comment by bstifle** [comment]
@nick-esri @TheBlueDog couple more iterations without changing the reds.

if we make the text smaller we need to consider Small which would then use 10px. i think those look good with the medium weight.

not a fan of the grayer text for these, as the messages themselves tend to be important

![image](https://user-images.githubusercontent.com/14083530/114423190-8ab06c80-9b7c-11eb-83cf-98c6b85dddec.png)

**Comment by nick-esri** [comment]
Thanks for these iterations @bstifle. I'm with you on the grey- now that I see them all together. 

I quite like the 12px version in the 3rd column. Making the error text smaller does bring an element of hierarchy to the component. 
<img width="441" alt="image" src="https://user-images.githubusercontent.com/3908500/114617780-95e7c300-9c76-11eb-9dea-047c99b8c9c7.png">

I do find the error icon slightly large compared to the text, and it might be worth tightening up the space between the icon and the error text.

**Comment by bstifle** [comment]
yeah man agreed on the space between. unfortunately we cant really make the icon smaller. it will make the lines fall off pixel and get all blurry unfortunately. i can def see a case for 12px icons but man with 633 concepts in the ui-icon library that could open a can of worms.

**Comment by rmstinson** [comment]
Had a note about regular Hint Texts under input fields - for anything that's non-info/warning/error, I think it would be better if it were styling with Regular Avenir rather than Medium. For example, a Medium character counter easily becomes one of the most emphasized objects in a form, which it really doesn't deserve IMO.

![Screen Shot 2021-04-14 at 4 01 31 PM](https://user-images.githubusercontent.com/12650705/114791111-c1cc7c80-9d3a-11eb-8814-682ffe9a5537.png)

**Comment by TheBlueDog** [comment]
Real-world example of what a form could look like in our app. The text below the input being a heavier weight than the label can be quite distracting and doesn't fall in like with the information hierarchy.

![Screen Shot 2021-04-20 at 12 46 28 PM](https://user-images.githubusercontent.com/5419306/115434227-769ce880-a1d6-11eb-864e-09ca447848b1.png)

**Comment by bstifle** [comment]
@rmstinson 100% agree hint should be regular. i think hint should ALSO be ui-text-3 to be more subdued

@TheBlueDog the text would be then styled identically to the labels. isn't the idea for errors to be more prominent since they require immediate attention?

in general, i kinda like the idea of keeping things medium, and dropping them down one font size for error per this thing
![image](https://user-images.githubusercontent.com/14083530/115462794-b1ac1580-a1f0-11eb-8e3c-2d32f12f0f86.png)

**Comment by bstifle** [comment]
another pass, based on feedback from adam.

Would be pretty smart to keep hint in a separate slot, would also be read directly after label for screenreaders.

and the hint can persist when a status message is displayed. @rmstinson @TheBlueDog @nick-esri thoughts?

![image](https://user-images.githubusercontent.com/14083530/115595229-4b7ccc80-a29c-11eb-96c2-9114c603231a.png)

**Comment by macandcheese** [comment]
For the text area character count and optional (or translated optional string), I think those are two areas we can look at to standardize around as well.

**Comment by bstifle** [comment]
yeah agreed. some more considerations.

![image](https://user-images.githubusercontent.com/14083530/115607854-1af05f00-a2ab-11eb-9479-2a1dbb2e58de.png)

**Comment by bstifle** [comment]
All here, please take a look at this semi-exhaustive list of visuals for validation:

![Status Validation](https://user-images.githubusercontent.com/14083530/127203472-7ca640ad-7879-413b-a511-1b1bd04fe844.png)

Key features:

- Space between validation icons and text is 8px in all scales
- validation text is 1 font-size smaller than the label of it's parent in all scales
- validation text is ui-text-1 (#151515) and weight is medium
- validation on scale small has 4px top and bottom margin
- validation on scale medium and large have 8px top and bottom margin
- hint text is 1 font-size smaller than the label of it's parent in all scales
- hint text is ui-text-3 and regular weight


Nothing here on this list has changed since April 21. Just tryna get it all on paper.

Gonna leave this up for design discussion until Friday July 27

**Comment by bstifle** [comment]
Closing design discussion, moving over to triage

**Comment by bstifle** [comment]
@justinhough this is the most up do date and accurate comment:

https://github.com/Esri/calcite-components/issues/1592#issuecomment-887715401
